File: src/AutobotsPlaywrightFramework/PageObjects/Neonatal/PersonalDetailsFormPlaywright.cs
using AutobotsPlaywrightFramework.Extensions;
using AutobotsPlaywrightFramework.Models.Neonatal;
using Microsoft.Playwright;

namespace AutobotsPlaywrightFramework.PageObjects.Neonatal;

/// <summary>
/// Playwright page object for the Personal Details form.
/// </summary>
public sealed class PersonalDetailsFormPlaywright
{
    private readonly IPage _page;

    private const string FirstNameTextBox = "#firstName";
    private const string SurnameTextBox = "#surname";
    private const string EmailTextBox = "#email";
    private const string GenderFrame = "iframe#genderFrame";
    private const string MaleOption = "#genderMale";
    private const string ContinueButton = "#continue";

    /// <summary>
    /// Initializes a new instance of the <see cref="PersonalDetailsFormPlaywright"/> class.
    /// </summary>
    public PersonalDetailsFormPlaywright(IPage page)
    {
        _page = page;
    }

    /// <summary>
    /// Fills employee personal details.
    /// </summary>
    public async Task FillAsync(EmployeeDetails employee)
    {
        await (await _page.WaitUntilVisible(FirstNameTextBox).ConfigureAwait(false)).FillAsync(employee.FirstName).ConfigureAwait(false);
        await (await _page.WaitUntilVisible(SurnameTextBox).ConfigureAwait(false)).FillAsync(employee.Surname).ConfigureAwait(false);
        await (await _page.WaitUntilVisible(EmailTextBox).ConfigureAwait(false)).FillAsync(employee.Email).ConfigureAwait(false);
        await _page.FrameLocator(GenderFrame).Locator(MaleOption).ClickAsync().ConfigureAwait(false);
    }

    /// <summary>
    /// Submits the form and accepts the confirmation alert.
    /// </summary>
    public Task SubmitAndAcceptAlertAsync()
    {
        return _page.WaitForAlertAndAccept(() => _page.Locator(ContinueButton).ClickAsync());
    }
}

File: src/pageObjects/neonatal/personalDetailsFormPlaywright.ts
import { Page } from '@playwright/test';
import { waitForAlertAndAccept, waitUntilVisible } from '../../extensions/playwrightExtensions';
import { EmployeeDetails } from '../../models/neonatal/employeeBookingData';

/** Playwright page object for Personal Details form. */
export class PersonalDetailsFormPlaywright {
  private readonly firstName = '#firstName';
  private readonly surname = '#surname';
  private readonly email = '#email';
  private readonly genderFrame = 'iframe#genderFrame';
  private readonly maleOption = '#genderMale';
  private readonly continueButton = '#continue';

  public constructor(private readonly page: Page) {}

  public async fill(employee: EmployeeDetails): Promise<void> {
    await (await waitUntilVisible(this.page, this.firstName)).fill(employee.firstName);
    await (await waitUntilVisible(this.page, this.surname)).fill(employee.surname);
    await (await waitUntilVisible(this.page, this.email)).fill(employee.email);
    await this.page.frameLocator(this.genderFrame).locator(this.maleOption).click();
  }

  public async submitAndValidateAlert(expectedMessages: readonly string[]): Promise<void> {
    const message = await waitForAlertAndAccept(this.page, () => this.page.locator(this.continueButton).click());
    const matched = matchesExpectedAlert(message, expectedMessages);
    if (!matched) {
      throw new Error(`Alert message '${message}' does not match expected values: ${expectedMessages.join(', ')}`);
    }
  }
}

/** Returns true when dialog message contains one of the expected values (case-insensitive). */
export function matchesExpectedAlert(actualMessage: string, expectedMessages: readonly string[]): boolean {
  return expectedMessages.some((expected) => actualMessage.toLowerCase().includes(expected.toLowerCase()));
}
